src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokenInfoService.cs

```diff
@@ -57,9 +87,103 @@ private static IReadOnlyList<SyntaxResult> VisitAllNodes(RazorCodeDocument razor
                 previousResult = result;
             }
 
-            return new SemanticTokens
+            var resultId = Guid.NewGuid();
+
+            var tokensResult = new SemanticTokens
             {
-                Data = data
+                Data = data.ToArray(),
+                ResultId = resultId.ToString()
+            };
+
+            _semanticTokenCache.Set(resultId.ToString(), data);
+
+            return tokensResult;
+        }
+
+        private static SemanticTokensOrSemanticTokensEdits ConvertSyntaxTokensToSemanticEdits(
+            IReadOnlyList<SyntaxResult> syntaxResults,
+            IReadOnlyList<uint> previousResults,
+            RazorCodeDocument codeDocument)
+        {
+            var newTokens = ConvertSyntaxTokensToSemanticTokens(syntaxResults, codeDocument);
+            var oldData = previousResults;
+
+            if (oldData is null || oldData.Count == 0)
+            {
+                return newTokens;
+            }
+
+            // The below algorithm was taken from OmniSharp/csharp-language-server-protocol at
+            // https://github.com/OmniSharp/csharp-language-server-protocol/blob/bdec4c73240be52fbb25a81f6ad7d409f77b5215/src/Protocol/Document/Proposals/SemanticTokensDocument.cs#L156
```

Comment:
❤️  

I believe we need to do some sort of component governance piece since we're adopting this. Also, if we want this method to be purely based off of the implementation there I'd recommend pulling it into its own file and leaving this comment there so the "copy" can be isolated.